src/client.rs

```diff
@@ -0,0 +1,190 @@
+pub struct Client<T = reqwest::Client, A = any::Credentials> {
+    inner: T,
+    credentials: A,
+}
+
+pub type DefaultClient = Client<reqwest::Client, Option<Arc<any::Credentials>>>;
+
+impl<T: Execute + fmt::Debug, A: Authorization + fmt::Debug> Client<T, A> {
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    pub fn new(inner: T, credentials: A) -> Self {
+        Self { inner, credentials }
+    }
+
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    pub fn set_credentials(&mut self, credentials: A) {
+        self.credentials = credentials;
+    }
+
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    pub fn with_credentials<B>(self, credentials: B) -> Client<T, B>
+    where
+        B: Authorization + fmt::Debug,
+    {
+        Client::new(self.inner, credentials)
+    }
+
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    pub fn inner(&self) -> &T {
+        &self.inner
+    }
+}
+
+impl<T, A> Default for Client<T, A>
+where
+    T: Execute + Default + fmt::Debug,
+    A: Authorization + Default + fmt::Debug,
+{
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    fn default() -> Self {
+        Self::new(T::default(), A::default())
+    }
+}
+
+impl<T, A> From<A> for Client<T, A>
+where
+    T: Execute + Default + fmt::Debug,
+    A: Authorization + fmt::Debug,
+{
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    fn from(credentials: A) -> Self {
+        Self::new(T::default(), credentials)
+    }
+}
+
+impl<T> From<T> for Client<T, ()>
+where
+    T: Execute + fmt::Debug,
+{
+    #[cfg_attr(feature = "tracing", tracing::instrument)]
+    fn from(client: T) -> Self {
+        Self::new(client, ())
+    }
+}
+
+impl<T, A> Execute for Client<T, A>
+where
+    T: Execute<Error: Send> + Send + Sync + 'static,
+    A: Authorization<Error: Send> + Send + Sync + 'static,
+    Self: Clone,
+{
+    type Error = ClientError<<T as Execute>::Error, <A as Authorization>::Error>;
+
+    #[allow(clippy::cast_precision_loss)]
+    fn execute(
+        &self,
+        mut request: reqwest::Request,
+    ) -> impl Future<Output = Result<Response, Self::Error>> + Send + 'static {
+        let this = self.clone();
+
+        async move {
+            if !request.headers().contains_key(header::AUTHORIZATION) {
+                if let Some(authorization) = this
+                    .credentials
+                    .authorization(&request)
+                    .map_err(ClientError::Authorization)?
+                {
+                    let val = {
+                        let mut val = authorization.parse::<HeaderValue>()?;
+                        val.set_sensitive(true);
+                        val
+                    };
+                    request
+                        .headers_mut()
+                        .try_insert(header::AUTHORIZATION, val)?;
+                }
+            }
+
+            #[cfg(any(feature = "logging", feature = "metrics"))]
```

Comment:
Should we move logging and metrics to the `reqwest::Client` implementation of `Execute` trait ?